### PR TITLE
poc: Byte source trait

### PR DESCRIPTION
**Pull request by nameexhaustion**
POC for a byte source trait to allow a unified method to perform file reads from local and cloud.

Currently I plan to use this for the new streaming engine.

### CODE REVIEW

File: crates/polars-stream/src/utils/byte_source.rs [hunk 1]
```diff
@@ -0,0 +1,329 @@
...
+    #[derive(Clone)]
+    #[allow(unused)]
+    pub(super) enum BytesKeepAlive {
+        Local(Arc<memmap::Mmap>),
+        Cloud(bytes::Bytes),
+    }
+
+    /// TODO: Add slicing to owned data
+    #[derive(Clone)]
+    pub struct Bytes {
```
Comment (nameexhaustion) [inline crates/polars-stream/src/utils/byte_source.rs:26]: This is the `Bytes` type that is returned from a `ByteSource`

File: crates/polars-stream/src/utils/byte_source.rs [hunk 2]
```diff
@@ -0,0 +1,329 @@
...
+                    .try_build_from_path("/Users/nxs/git/polars/.env/iris.parquet", None)
+                    .await
+                    .unwrap(),
+                &async_executor::ExecutorScope::new(),
+            )
+            .await;
+
+            dbg!(local_result.map(|x| x.num_rows));
+
+            let cloud_result = read_parquet_metadata(
```
Comment (nameexhaustion) [inline crates/polars-stream/src/utils/byte_source.rs:240]: `DynByteSource` lets us use the same `read_parquet_metadata` function to read from both local and cloud

File: crates/polars-stream/src/utils/byte_source.rs [hunk 3]
```diff
@@ -0,0 +1,329 @@
...
+
+mod bytes_private {
+    use std::sync::Arc;
+
+    /// This is for keeping the buffers alive.
+    #[derive(Clone)]
+    #[allow(unused)]
+    pub(super) enum BytesKeepAlive {
+        Local(Arc<memmap::Mmap>),
+        Cloud(bytes::Bytes),
```
Comment (coastalwhite) [inline crates/polars-stream/src/utils/byte_source.rs:21]: This is basically the same as `polars_utils::mmap::MmapSlice`. I think you should use that.

File: crates/polars-stream/src/utils/byte_source.rs [hunk 4]
```diff
@@ -0,0 +1,329 @@
...
+        _cloud_options: Option<&CloudOptions>,
+    ) -> PolarsResult<Self> {
+        let file = Arc::new(
+            tokio::fs::File::open(path)
+                .await
+                .map_err(|err| _limit_path_len_io_err(path.as_ref(), err))?
+                .into_std()
+                .await,
+        );
+        let mmap = Arc::new(unsafe { memmap::Mmap::map(file.as_ref()) }.map_err(to_compute_err)?);
```
Comment (coastalwhite) [inline crates/polars-stream/src/utils/byte_source.rs:77]: As far as I am aware and what Orson told me, we should not be memory mapping files for the streaming engine. The Parquet reader takes heap allocated data for this reason.

File: crates/polars-stream/src/utils/byte_source.rs [hunk 5]
```diff
@@ -0,0 +1,329 @@
...
+
+    async fn get_size(&self) -> PolarsResult<usize> {
+        Ok(self.mmap.as_ref().len())
+    }
+
+    async fn get_range(&self, range: Range<usize>) -> PolarsResult<Bytes> {
+        // TODO: Change from mmap to file-read - this currently doesn't trigger any reads.
+        let slice = unsafe { std::mem::transmute(self.mmap.as_ref().as_ref().get(range).unwrap()) };
+
+        Ok(Bytes::new(slice, BytesKeepAlive::Local(self.mmap.clone())))
```
Comment (coastalwhite) [inline crates/polars-stream/src/utils/byte_source.rs:90]: This should probably use the parquet::MemReader or parquet::MemReaderSlice to ensure that it is heap-allocated and can be passed to the parquet reader.

**Pull request closed without merging**